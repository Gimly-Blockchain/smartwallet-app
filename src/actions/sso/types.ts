import { PublicProfileClaimMetadata } from 'cred-types-jolocom-core/types'

/**
 * @dev Simply using all claims required by the public profile
 */

export type IssuerPublicProfileSummary = PublicProfileClaimMetadata['claimInterface']

/**
 * @dev An identity summary is composed of a DID + all public info (currently public profile)
 */
export interface IdentitySummary {
  did: string
  publicProfile?: IssuerPublicProfileSummary
}

export interface AuthenticationRequestSummary {
  callbackURL: string
  requester: IdentitySummary
  description: string
  requestJWT: string
}