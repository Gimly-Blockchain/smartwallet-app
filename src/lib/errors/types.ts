import { routeList } from 'src/routeList'
import { ErrorCode } from './codes'

export { ErrorCode }

export interface IAppError {
  origError: any
  navigateTo: routeList
}

export interface UserReport {
  userError: string | undefined
  userDescription: string
  userContact: string
  sendPrivateData: boolean
}

export interface ErrorReport extends UserReport {
  error: IAppError | Error | undefined
}
