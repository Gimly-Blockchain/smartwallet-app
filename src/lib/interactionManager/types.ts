
import { JSONWebToken, JWTEncodable } from 'jolocom-lib/js/interactionTokens/JSONWebToken'


/* Responsabilities:
 * 1. continue an interaction
 */
export interface IInteractionManager {
    tokenRecieved: (token: JSONWebToken<JWTEncodable>) => void

}
