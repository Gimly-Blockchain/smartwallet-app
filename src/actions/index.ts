import * as accountActions from 'src/actions/account/'
import * as registrationActions from 'src/actions/registration/'
import * as navigationActions from 'src/actions/navigation/'
import * as genericActions from 'src/actions/generic/'
import * as ssoActions from 'src/actions/sso/'
import * as interactionHandlerActions from 'src/actions/interactionHandler'
import * as paymentActions from 'src/actions/payment/'

export {
  accountActions,
  registrationActions,
  navigationActions,
  genericActions,
  ssoActions,
  interactionHandlerActions,
  paymentActions
}
