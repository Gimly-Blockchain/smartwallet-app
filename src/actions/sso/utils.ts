import { IssuerPublicProfileSummary } from './types'
import { Identity } from 'jolocom-lib/js/identity/identity'

export const parsePublicProfile = (identity: Identity) => {
  const { publicProfile, did } = identity
  if (!publicProfile) {
    return {
      did,
    }
  }
  const { id, ...parsedProfile } = publicProfile.claim
  return {
    did,
    publicProfile: parsedProfile as IssuerPublicProfileSummary,
  }
}
