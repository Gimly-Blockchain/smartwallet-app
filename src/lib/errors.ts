import {routeList} from '../routeList'
import strings from '../locales/strings'
import ErrorCode from './errorCodes'

export class AppError extends Error {
  // private code: ErrorCode
  public origError: any
  navigateTo: routeList

  public constructor(code = ErrorCode.Unknown, origError?: any, navigateTo: routeList = routeList.Home) {
    super(strings[code] || strings[ErrorCode.Unknown])
    // this.code = code
    this.origError = origError
    this.navigateTo = navigateTo
  }
}

export const errorTitleMessages = [strings.DAMN, strings.OH_NO, strings.UH_OH]
