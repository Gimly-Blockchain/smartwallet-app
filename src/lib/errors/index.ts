import { routeList } from 'src/routeList'
import strings from '../../locales/strings'
import { ErrorCode } from '@jolocom/sdk/js/src/lib/errors/types'
export { ErrorCode }

export interface IAppError {
  origError: any
  navigateTo: routeList
}
export interface UserReport {
  userError: string | undefined
  userDescription: string
  userContact: string
  sendPrivateData: boolean
}

export interface ErrorReport extends UserReport {
  error: IAppError | Error | undefined
}
import { reportErrorToSentry, initSentry } from 'src/lib/errors/sentry'

export class AppError extends Error implements IAppError {
  // private code: ErrorCode
  public origError: any
  public navigateTo: routeList

  public constructor(
    code = ErrorCode.Unknown,
    origError?: any,
    navigateTo: routeList = routeList.AppInit,
  ) {
    super(strings[code] || strings[ErrorCode.Unknown])
    // this.code = code
    this.origError = origError
    this.navigateTo = navigateTo
  }
}

export const errorTitleMessages = [strings.DAMN, strings.OH_NO, strings.UH_OH]

export function reportError(report: ErrorReport) {
  let extraData
  if (report.error instanceof AppError && report.error.origError) {
    extraData = {
      AppError: report.error.message,
      UserError: report.userError,
      UserDescription: report.userDescription,
      UserContact: report.userContact,
      sendPrivateData: report.sendPrivateData,
    }
    report.error = report.error.origError
  }
  return reportErrorToSentry(report, extraData)
}

export const initErrorReporting = initSentry
