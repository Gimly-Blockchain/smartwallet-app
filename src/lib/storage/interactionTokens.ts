import { JSONWebToken } from 'jolocom-lib/js/interactionTokens/JSONWebToken'
import { InteractionType } from 'jolocom-lib/js/interactionTokens/types'
import { consumeCredentialRequest } from '../../actions/sso'
import { consumeAuthenticationRequest } from '../../actions/sso/authenticationRequest'
import { consumeCredentialOfferRequest } from '../../actions/sso/credentialOfferRequest'
import { Authentication } from 'jolocom-lib/js/interactionTokens/authentication'
import { CredentialOfferRequest } from 'jolocom-lib/js/interactionTokens/credentialOfferRequest'
import { CredentialRequest } from 'jolocom-lib/js/interactionTokens/credentialRequest'
import { PaymentRequest } from 'jolocom-lib/js/interactionTokens/paymentRequest'
import { consumePaymentRequest } from '../../actions/sso/paymentRequest'
import { SendResponse } from 'src/lib/transportLayers'
/**
 * @param Metadata should not need to be passed to credential receive because it comes from cred Offer
 * Furthermore, this only needs to be defined for requests
 */

export const interactionHandlers = {
  [InteractionType.Authentication]: <T extends JSONWebToken<Authentication>>(
    interactionToken: T,
    send: SendResponse,
  ) => consumeAuthenticationRequest(interactionToken, send),
  [InteractionType.CredentialRequest]: <T extends JSONWebToken<CredentialRequest>>(
    interactionToken: T,
    send: SendResponse,
  ) => consumeCredentialRequest(interactionToken, send),
  [InteractionType.CredentialOfferRequest]: <
    T extends JSONWebToken<CredentialOfferRequest>
  >(
    interactionToken: T,
    isDeepLinkInteraction: boolean,
  ) => consumeCredentialOfferRequest(interactionToken, isDeepLinkInteraction),
  [InteractionType.PaymentRequest]: <T extends JSONWebToken<PaymentRequest>>(
    interactionToken: T,
    isDeepLinkInteraction: boolean,
  ) => consumePaymentRequest(interactionToken, isDeepLinkInteraction),
}
