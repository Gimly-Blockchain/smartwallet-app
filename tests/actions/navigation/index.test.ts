import { navigationActions, interactionHandlers } from '../../../src/actions'
import { JolocomLib } from 'jolocom-lib'
import { JSONWebToken } from 'jolocom-lib/js/interactionTokens/JSONWebToken'
import { InteractionType } from 'jolocom-lib/js/interactionTokens/types'
import { createMockStore } from 'tests/utils'
import { AppError } from 'src/lib/errors'

describe('Navigation action creators', () => {
  describe('handleDeepLink', () => {
    const jwt = 'mockJWT'

    const mockDid = 'did:jolo:mock'

    const mockStore = createMockStore(
      {
        account: {
          did: {
            did: mockDid,
          },
        },
        generic: {
          appWrapConfig: {},
        },
      },
      {
        passwordStore: {
          getPassword: jest.fn().mockResolvedValue('secret123'),
        },
        identityWallet: {
          validateJWT: jest.fn().mockResolvedValue(true),
        },
      },
    )

    const parseInteractionTokenSpy = jest.spyOn(
      JolocomLib.parse.interactionToken,
      'fromJWT',
    )

    beforeEach(() => {
      const interactionHandlersSpies = {}
      Object.keys(interactionHandlers).forEach(typ => {
        interactionHandlersSpies[typ] = jest
          // @ts-ignore bleh
          .spyOn(interactionHandlers, typ)
          // @ts-ignore bleh
          .mockReturnValue({ type: `MOCK_${typ}_INTERACTION_TOKEN_HANDLER` })
      })

      jest.useFakeTimers()
      mockStore.reset()
      parseInteractionTokenSpy.mockClear()
    })

    // TODO: refactor test case to account for identity check when deeplinking
    it('should extract the route name and param from the URL', async () => {
      parseInteractionTokenSpy.mockImplementation(jwt => {
        const token = new JSONWebToken()
        token.payload = { typ: InteractionType.CredentialRequest }
        return token
      })

      const action = navigationActions.handleDeepLink(
        'jolocomwallet://consent/' + jwt,
      )

      await mockStore.dispatch(action)
      jest.runAllTimers()
      expect(mockStore.getActions()).toMatchSnapshot()
      expect(parseInteractionTokenSpy).toHaveBeenCalledWith(jwt)
    })

    it('should throw AppError if route was not correct', () => {
      const action = navigationActions.handleDeepLink(
        'jolocomwallet://somethingElse/' + jwt,
      )

      expect(() => mockStore.dispatch(action)).toThrow(AppError)
      expect(mockStore.getActions()).toMatchSnapshot()
      expect(parseInteractionTokenSpy).not.toHaveBeenCalled()
    })
  })
})
