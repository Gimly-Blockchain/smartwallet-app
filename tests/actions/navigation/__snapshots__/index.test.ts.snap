// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation action creators handleDeepLink should extract the route name and param from the URL 1`] = `
Array [
  Object {
    "type": "APPWRAP_SHOW_LOADER",
    "value": true,
  },
  Object {
    "type": "MOCK_credentialRequest_INTERACTION_TOKEN_HANDLER",
  },
]
`;

exports[`Navigation action creators handleDeepLink should throw AppError if route was not correct 1`] = `Array []`;
